src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -593,7 +593,7 @@ private bool CheckAttributeTypeViability(Symbol symbol, bool diagnose, ref Diagn
             if (symbol.Kind == SymbolKind.NamedType)
             {
                 var namedType = (NamedTypeSymbol)symbol;
-                if (namedType.IsGenericType)
+                if (namedType.IsGenericType && !Compilation.LanguageVersion.AllowGenericAttributes())
```

Comment:
I think we have a problem here.
Having `CheckAttributeTypeViability` return `false` depending on LangVer here affects the callers (such as `LookupAttributeType`). But a general rule of the compiler is that we bind the same regardless of LangVer (we only produce different diagnostics).
Look at `LookupAttributeType` and see if we can produce that problem.
I suspect we need to remove this `IsGeneric` check here (it is a valid attribute type, regardless of LangVer, and we'll add a diagnostic elsewhere (in C# 8, we disallow that attribute, even though it is considered viable).

We'll need to test scenarios with ambiguities. For example we have a type `Foo<T>` that's not an attribute and a type `FooAttribute` that is, then we write `[Foo]`. We'll need to maintain the current behavior (whatever that is).

I need to read the spec on this...

Reply:
Agreed. Generally speaking, in order to follow our current design preference that we bind the same way but just change diagnostics when that is possible:
(1) We should now unconditionally bind to a generic attribute *if it has the right number of type arguments* at the use site. This method appears to classify viability independent of arity, which looks like it may affect binding.
(2) Then, if it was generic and the language version doesn't support it, produce an error. This could technically be a syntax-only check (since the type arguments appear in source) but better in semantics.

And agreed, we need to ensure we have tests that demonstrate that we have not changed behavior for existing code that uses nongeneric attributes (especially since we do the lookup twice).


---
In reply to: [414214253](https://github.com/dotnet/roslyn/pull/26337#discussion_r414214253) [](ancestors = 414214253)

Reply:
@gafter maybe I don't understand something
If the check is syntax only what happen in this case?
```c#
class FooAttribute<T> : Attribute {}
[Foo]
class Program
{
}
```
According to syntax I cant know that Foo is generic type

Reply:
also using alias case need the bind phase

Reply:
You're correct that looking at the syntax `[Type]`, one cannot obviously tell whether the type is generic. To do that we lookup the type and evaluate the candidate symbols (symbols have a well-defined arity).

Reply:
Quick update: we briefly discussed this feature in LDM today. We moved it to be a candidate for next release (C# 10) and there is still interest, assuming we can avoid introducing any compat issue. 
Let us know if you're still up for moving this forward. Thanks! :-)

Reply:
@jcouv  I'm always here and waiting for a clear instructions how to finish this feature (as done in the coreclr part)
I understand that you have time restriction and other commitments
And happy it's finally targeted to C# 10
I hope now the progress will be faster

Reply:
@AviAvni I didn't realize you're waiting for instructions. From earlier discussion, we've spotted a problem with making viability depend on language version. The first step I recommended above was to seek a scenario that highlights the problem.